### ISSUE TITLE
New Makefile Output & Debugging

### ISSUE THREAD

**Issue by metaskills** [aws/aws-sam-cli#2006]
First, thanks a ton for the new 0.51.0 Makefile support! This is going to save a bunch of time on Rails & Lambda build times since we typically have to build CSS/JS assets using non-deployment gems. Hence having full control allows our CI/CD (GitHub Actions) to cache different dev/deploy gem vendor directories.

That said, I am new to Makefiles and found their adoption to be a little difficult. I found myself wanting to `echo` parts of the build but nothing was showing during `sam build`. Is there something this library can do (or myself) when using a Makefile to see what is happing in my own files? 

I also realize this just might be having to learn make. But thought I would ask first. Thanks again!

**Comment by smacintyre** [comment]
When creating your Makefile it will be easier to initially run make directly while you get things setup. That way you can echo things and see what is going on. Also, keep in mind that make is going to be called for each function with `Metadata: makefile` set.

You can run the build for a function by running following. Where `FUNCTION` is the logical name of your function and `PATH` is where you want the files to be deposited for your testing.. Then you can see the full output of your function including any echos. By default commands are executed with `/bin/sh`.

```bash
make build-[FUNCTION] ARTIFACTS_DIR=[PATH]
```

Make is very flexible and worth learning. A quick tutorial is at http://makefiletutorial.com/ . Also check out the [GNU Make Standard Library](https://gmsl.sourceforge.io/). Once you have it setup you shouldn't have to modify to much.

**Comment by metaskills** [comment]
Thanks, I'd like to propose that maybe someone should not have to run the Makefile manually to debug it. I noticed that make.py here (https://github.com/awslabs/aws-lambda-builders/blob/develop/aws_lambda_builders/workflows/custom_make/make.py#L85-L90) does return the output. However, the make action does not use it (https://github.com/awslabs/aws-lambda-builders/blob/develop/aws_lambda_builders/workflows/custom_make/actions.py#L87)

Maybe it should is what I am asking?

**Comment by sriram-mv** [comment]
As @smacintyre pointed out, yeah the intention behind the usage of the makefile was so that users can use it to adjust everything as required, before being picked up by sam build.

However, that being said. Streaming the output of triggering the make - build target, seems reasonable.

**Comment by metaskills** [comment]
Gonna close this.

**Comment by joshwand** [comment]
This is especially relevant when using `--use-container` to run the makefile inside the container. Not nearly so easy to mess around with the makefile in that scenario.

**Comment by ricky-at-picnic** [comment]
I'm now trying to build a Makefile and wanting to see output. What is the status on this?

**Comment by diogoko** [comment]
To see the output of the commands in case of failure of a Makefile rule, you can redirect all STDOUT to STDERR using `1>&2`:

```makefile
my-rule:
	npm run build-my-lambda 1>&2
```

**Comment by maschnetwork** [comment]
Any updates here? Especially for GraalVM native image build which take a few minutes you will only get the output after that time with no option to check the current progress.

We are using --use-container with a custom build image and a Makefile.

**Comment by mndeveci** [comment]
Hi there,

This is been fixed since [SAM CLI v1.62.0](https://github.com/aws/aws-sam-cli/releases/tag/v1.62.0) with this PR from AWS Lambda Builders https://github.com/aws/aws-lambda-builders/pull/399
